### ISSUE TITLE
change App project to DLL problems

### ISSUE THREAD

**Issue by DigitalPilgrim** [dotnet/wpf#2812]
* .NET Core Version: 3.1
* Visual Studio 2019 Version 16.5.0 (new / fresh instalation and new project)
after update to 16.5.1 same problems

 **Description:**
I have WPF NetFramework app converted in to DLL where I used App.Xaml. Now I am trying to do same with WPF Net Core.

---

**1.  Problem**
 I can not change App.Xaml to Page from Properties. I got this error:

> Property Value is not valid.
> An error has occurred while saving the edited properties listed below:
> Build Action
> One or more values are invalid. Only included item inside the project can be moved.
> Parameter name: currentItem

same problem when I change Window from Page to Content and back

> An error has occurred while saving the edited properties listed below:
> Build Action
> One or more values are invalid. Cannot add 'MainWindow.xaml' to the project, because the path is explicitly excluded from the project (C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.NET.Sdk.WindowsDesktop.targets (35,5)).

---

**2.  Problem**
when WPF app is converted to ClassLibrary (DLL) then Windows and Application can not use InitializeComponent() function. And when I try build it I got this error:

> SeverityCodeDescriptionProjectFileLineSuppression State ErrorCS0103The name 'InitializeComponent' does not exist in the current contextWpfApp1CoreE:\PathToMyProject\WpfApp1Core\WpfApp1Core\MainWindow.xaml.cs25Active

---

**3.  Problem**
I created another project WPF Net.Core app and InitializeComponent working, but error when changing properties for Xaml is still here. When I try compile it, I got this errors:

>  1>C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.WinFX.targets(225,9): error MC1002: Library project file cannot specify ApplicationDefinition element. 1>C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.WinFX.targets(225,9): error BG1003: The project file contains a property value that is not valid.

> SeverityCodeDescriptionProjectFileLineSuppression State ErrorNETSDK1022Duplicate 'Page' items were included. The .NET SDK includes 'Page' items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultPageItems' property to 'false' if you want to explicitly include them in your project file. For more information, see https://aka.ms/sdkimplicititems. The duplicate items were: 'App.xaml'WpfApp2CoreC:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.NET.Sdk.WindowsDesktop.targets69

> SeverityCodeDescriptionProjectFileLineSuppression State ErrorNETSDK1022Duplicate 'Page' items were included. The .NET SDK includes 'Page' items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultPageItems' property to 'false' if you want to explicitly include them in your project file. For more information, see https://aka.ms/sdkimplicititems. The duplicate items were: 'MainWindow.xaml'WpfApp2CoreC:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.NET.Sdk.WindowsDesktop.targets69

---
btw. I posted it [there](https://developercommunity.visualstudio.com/content/problem/958304/wpf-core-app-change-to-dll-problems.html), but I was redirected to post it here.

**Comment by vatsan-madhavan** [comment]
Consider posting your project along with [binary build logs](https://aka.ms/msbuild/binary-log) - it makes it so much easier to troubleshoot. 

There is a very good chance that you'll see the same problems when building with `dotnet `or `msbuild` , and generating msbuild binary logs from commandline is much easier (tag along `/bl` to the build command).

**Comment by vatsan-madhavan** [comment]
> **1. Problem**
> I can not change App.Xaml to Page from Properties. I got this error:
> 
> > Property Value is not valid.
> > An error has occurred while saving the edited properties listed below:
> > Build Action
> > One or more values are invalid. Only included item inside the project can be moved.
> > Parameter name: currentItem
> 
> same problem when I change Window from Page to Content and back
> 
> > An error has occurred while saving the edited properties listed below:
> > Build Action
> > One or more values are invalid. Cannot add 'MainWindow.xaml' to the project, because the path is explicitly excluded from the project (C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.NET.Sdk.WindowsDesktop.targets (35,5)).

In an SDK style WindowsDesktop project, 

- *.xaml are all treated as `Page` items by default, unless they are named `App.xaml` (C#) or `Application.xaml` (VB). 
- App.xaml/Application.xaml is treated as `ApplicationDefinition` by default. 

To change this default behavior, you have to override the _implicit_ behavior. 

#### `ApplicationDefinition`

```xml
<!-- 
  Disables automatic globbing of App.xaml/Application.xaml into ApplicationDefinition item
  App.xaml/Application.xaml would now get treated as any other *.xaml file (typically, a Page)
  --> 
<PropertyGroup>
   <EnableDefaultApplicationDefinition>false</EnableDefaultApplicationDefinition>
</PropertyGroup>
```

#### `Page`
```xml
  <!-- 
    Disables automatic globbing of *.xaml files into Page items
    In order to include xaml files into Page items, they would have to be specified 
    in the project directly like this: 
    
      <ItemGroup>
        <Page Include="SomePage.xaml" />
      </ItemGroup>
  -->
  <PropertyGroup>
    <EnableDefaultPageItems>false</EnableDefaultPageItems>
  </PropertyGroup>
```


To achieve what you're trying to do, I suspect that you'll have to disable both `ApplicationDefinition` and `Page` implict-globbing, and specify everything in the project directly by hand. That way, you can decide exactly what ought to be `Content` and what would be `Page`. 

There is also a middle ground that is likely doable. If you just disable `ApplicationDefinition` implicit-globbing, then you can very likely use the VS IDE to set `App.xaml` as `Page`. Likewise, if you use the IDE to change another xaml (say, `Window.xaml`) to `content`, you'll likely end up with something like this in your project: 

```xml
  <PropertyGroup>
    <EnableDefaultApplicationDefinition>false</EnableDefaultApplicationDefinition>
  </PropertyGroup>
  
  <ItemGroup>
    <Page Remove="Window.xaml"  />
    <Content Include="Window.xaml" />
  </ItemGroup>
```

The `Page` items are still being implicitly globbed, which takes care of `App.xaml`. `Window.xaml` would also have been globbed implicitly into a `Page` item, which would have to be removed, and then added into the `Content` item to achieve your described end.

This should give you a general sense for how things work. If something doesn't work exactly as expected in the IDE, see if you can get it to work by editing the project yourself - that tends to work sometimes, and it would also help us differentiate between SDK bugs vs. an IDE bugs. 

For reference, you can see where this is all orchestrated:

- https://github.com/dotnet/wpf/blob/release/3.1/packaging/Microsoft.NET.Sdk.WindowsDesktop/targets/Microsoft.NET.Sdk.WindowsDesktop.props
- https://github.com/dotnet/wpf/blob/release/3.1/packaging/Microsoft.NET.Sdk.WindowsDesktop/targets/Microsoft.NET.Sdk.WindowsDesktop.targets

**Comment by vatsan-madhavan** [comment]
> 1>C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.WinFX.targets(225,9): error MC1002: Library project file cannot specify ApplicationDefinition element. 1>C:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.WinFX.targets(225,9): error BG1003: The project file contains a property value that is not valid.

I suspect that you are seeing this problem because you have a xaml file that matches the naming convention of `App.xaml` or `Application.xaml`, which then gets implicitly globbed into an `ApplicationDefinition` item. 

We could probably detect whether `$(OutputType)` is an `exe' or 'winexe' before globbing `ApplicationDefinition` (and when `$(OutputType)` is `library`, `module` etc. we'd just glob all XAML files into `Page` items.  /cc @rladuca, @SamBent, @dsplaisted 

That said, for now, your best bet is to leverage `EnableDefaultApplicationDefinition=false` if you are building a library that contains a XAML file with names like `App.xaml` or `Application.xaml`.

**Comment by vatsan-madhavan** [comment]
> SeverityCodeDescriptionProjectFileLineSuppression State ErrorNETSDK1022Duplicate 'Page' items were included. The .NET SDK includes 'Page' items from your project directory by default. You can either remove these items from your project file, or set the 'EnableDefaultPageItems' property to 'false' if you want to explicitly include them in your project file. For more information, see https://aka.ms/sdkimplicititems. The duplicate items were: 'App.xaml'WpfApp2CoreC:\Program Files\dotnet\sdk\3.1.200\Sdks\Microsoft.NET.Sdk.WindowsDesktop\targets\Microsoft.NET.Sdk.WindowsDesktop.targets69

The previous explanations have probably clarified for you what's going on here. If you had added something like this to your project, then the above error would have been inevitable:

```xml
  <ItemGroup>
    <Page Include="App.xaml" />
  </ItemGroup>
```

`App.xaml` was already being globbed implicitly by the WindowsDesktop SDK. Until you turn off this implicit behavior by leveraging either/or `EnableDefaultApplicationDefinition`, `EnableDefaultPageItems`, you could end up with the same XAML file being incorporated into the `Page` (or `ApplicationDefinition`) `ItemGroup` twice - the first time by the SDK itself and the second  time by you through the `ItemGroup` entry in the Project. Duplicate entries are not permitted - and thus we show an error.

**Comment by DigitalPilgrim** [comment]
Thank you very much, for detailed answer. It was very informative.
this solved all my problems
```
<PropertyGroup>
   <EnableDefaultApplicationDefinition>false</EnableDefaultApplicationDefinition>
</PropertyGroup>
```
This should by added automaticly when project is changed to ClassLibrary and otherwise, it should be deleted if project is changed back to application. I don't know where I can post this feature request?

**Comment by miloush** [comment]
@DigitalPilgrim I would be careful automatically changing the project file. If a developer overrides this setting then their override shouldn't be randomly changed.

It would make more sense to do it on explicit user action, for example when you choose Page in the Properties window for App.xaml: rather than giving you an error, the override should be set.

**Comment by DigitalPilgrim** [comment]
You are right. Maybe some checkbox under Properties for project will be helpfull, when project is changing in to ClassLibrary or Application.

**Comment by vatsan-madhavan** [comment]
How did you change your application to class-library? Did you change this by editing the project by hand, or is there a Visual Studio IDE option that is available that enabled you to point-and-click to make this change?

**Comment by DigitalPilgrim** [comment]
Yes, in Visual Studio IDE there is option to change project in to ClassLibrary from GUI. This is how I am doing it, when I want use App.Xaml in DLL, I create WPF App and change it into ClassLibrary by GUI from VS IDE. In .NET Framework it is only one click. 

![wpfOptions](https://user-images.githubusercontent.com/8824354/78541961-c2ad0300-77f6-11ea-8a56-ebf48b67dcff.jpg)

**Comment by dsplaisted** [comment]
I think it would make sense to default `EnableDefaultApplicationDefinition` to false for class libraries.  Possibly only when targeting .NET 5 and higher if we're concerned it could break existing projects.

**Comment by vatsan-madhavan** [comment]
I agree - definitely on .NET 5. I would also combine it with warnings for files with names like `App.xaml` or `Application.xaml` and call out that they are no longer being treated as `ApplicationDefinition` items. 

I'd suggest this should be done now for .NET Core 3.1 in servicing, before adoption becomes widespread and more confusion ensues. I don't really see how this would break anything. 

- WPF Class Library Projects either have `App.xaml` (or `Application.xaml`), or not. 
- If they do, then they **must** disable `ApplicationDefinition` globbing - otherwise it would lead to an error. Doing this on the developers behalf in the SDK is actually preferable. 
- If they don't, then disabling `ApplicationDefinition` globbing does no harm. Doing this on the WPF class-library projects' behalf in the SDK turns out to be benign. 

The benefits of fixing this in 3.1 are high IMO. These types of failures create a complex 'customer support' problem that's hard to diagnose, and hard to assist with. Either these cases would go undiagnosed and unsupported, or these will become expensive to support. 

Can anyone think of another way in which in it could pose a compat problem if this change were made in 3.1?

/cc @dotnet/wpf-developers
